File: mcp-server/nodejs/test-idempotency.js
import assert from "node:assert/strict";
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import { Client } from "@modelcontextprotocol/sdk/client/index.js";
import { StdioClientTransport } from "@modelcontextprotocol/sdk/client/stdio.js";

// Idempotency test for add_note — the Node mirror of ../src/test_idempotency.py.
// A retry is made safe two ways: a stable `idempotency_key`, or (with no key) an
// identical title+content. Either way: no duplicate, same id back.
//
// Run from the nodejs/ folder:  node test-idempotency.js

const dataRoot = fs.mkdtempSync(path.join(os.tmpdir(), "mcp-idem-"));

async function withClient(fn) {
  const t = new StdioClientTransport({
    command: "node",
    args: ["server.js"],
    env: { ...process.env, MCP_NOTES_DATA_ROOT: dataRoot },
  });
  const c = new Client({ name: "idem-test", version: "1.0.0" });
  await c.connect(t);
  try {
    return await fn(c);
  } finally {
    await c.close();
  }
}

const text = (r) => r.content[0].text;
const notesOf = async (c) => JSON.parse((await c.readResource({ uri: "notes://all" })).contents[0].text);

async function main() {
  await withClient(async (c) => {
    const start = (await notesOf(c)).length;

    // 1. No key: identical title+content twice -> one note.
    const r1 = await c.callTool({ name: "add_note", arguments: { title: "Deploy failed", content: "pod OOMKilled" } });
    const r2 = await c.callTool({ name: "add_note", arguments: { title: "Deploy failed", content: "pod OOMKilled" } });
    assert.match(text(r1), /added with id/, text(r1));
    assert.match(text(r2), /already stored/, `retry was not deduped: ${text(r2)}`);
    assert.equal((await notesOf(c)).length, start + 1, "identical retry created a second note");

    // 2. No key: same title, different content -> a real second note.
    const r3 = await c.callTool({ name: "add_note", arguments: { title: "Deploy failed", content: "different cause" } });
    assert.match(text(r3), /added with id/, text(r3));
    assert.equal((await notesOf(c)).length, start + 2);

    // 3. Explicit key: same key twice -> one note, first content wins, same id.
    const r4 = await c.callTool({
      name: "add_note",
      arguments: { title: "Retme", content: "first", idempotency_key: "op-42" },
    });
    const r5 = await c.callTool({
      name: "add_note",
      arguments: { title: "Retme", content: "SECOND try", idempotency_key: "op-42" },
    });
    const id4 = text(r4).trim().replace(/\.$/, "").split(" ").pop();
    assert.ok(text(r5).includes(id4), `same key returned a different note: ${text(r5)}`);
    assert.equal((await notesOf(c)).length, start + 3, "reused key created a second note");
    const stored = (await notesOf(c)).filter((n) => n.idempotencyKey === "op-42");
    assert.equal(stored.length, 1);
    assert.equal(stored[0].content, "first");
  });

  // 4. Survives a restart: a fresh process still dedups the keyed note.
  await withClient(async (c) => {
    const before = (await notesOf(c)).length;
    const r6 = await c.callTool({
      name: "add_note",
      arguments: { title: "x", content: "y", idempotency_key: "op-42" },
    });
    assert.match(text(r6), /already stored/, text(r6));
    assert.equal((await notesOf(c)).length, before);
  });

  console.log("PASS: add_note dedups on idempotency_key and on identical title+content.");
}

main()
  .then(() => {
    fs.rmSync(dataRoot, { recursive: true, force: true, maxRetries: 5, retryDelay: 100 });
    process.exit(0);
  })
  .catch((err) => {
    console.error("FAIL:", err);
    fs.rmSync(dataRoot, { recursive: true, force: true, maxRetries: 5, retryDelay: 100 });
    process.exit(1);
  });

File: mcp-server/src/test_idempotency.py
"""
Idempotency test for add_note.

add_note is a side-effecting operation and clients retry. Two ways a retry is
made safe:
  - pass a stable `idempotency_key`: the same key returns the note already
    stored under it;
  - pass no key: an identical title+content returns the existing note.
Either way, no duplicate row and the same id comes back.

Run from the mcp-server/ folder:
    python src/test_idempotency.py
"""

import asyncio
import json
import os
import sys
import tempfile
from pathlib import Path

from mcp import ClientSession, StdioServerParameters
from mcp.client.stdio import stdio_client
from mcp.types import ListRootsResult, Root


def _params(data_root: str) -> StdioServerParameters:
    return StdioServerParameters(
        command=sys.executable,
        args=["src/server.py"],
        env={**os.environ, "MCP_NOTES_DATA_ROOT": data_root},
    )


def _roots_cb(data_root: str):
    """add_note now requires the client to declare a root covering the store."""
    roots = [Root(uri=Path(data_root).resolve().as_uri())]

    async def cb(context) -> ListRootsResult:
        return ListRootsResult(roots=roots)

    return cb


async def _notes(session: ClientSession) -> list[dict]:
    result = await session.read_resource("notes://all")
    return json.loads(result.contents[0].text)


def _text(result) -> str:
    return result.content[0].text


async def main() -> None:
    with tempfile.TemporaryDirectory() as data_root:
        async with stdio_client(_params(data_root)) as (read, write):
            async with ClientSession(read, write, list_roots_callback=_roots_cb(data_root)) as session:
                await session.initialize()
                start = len(await _notes(session))

                # 1. No key: identical title+content twice -> one note.
                r1 = await session.call_tool("add_note", {"title": "Deploy failed", "content": "pod OOMKilled"})
                r2 = await session.call_tool("add_note", {"title": "Deploy failed", "content": "pod OOMKilled"})
                assert "added with id" in _text(r1), _text(r1)
                assert "already stored" in _text(r2), f"retry was not deduped: {_text(r2)}"
                assert len(await _notes(session)) == start + 1, "identical retry created a second note"

                # 2. No key: same title, different content -> a real second note.
                r3 = await session.call_tool("add_note", {"title": "Deploy failed", "content": "different cause"})
                assert "added with id" in _text(r3), _text(r3)
                assert len(await _notes(session)) == start + 2

                # 3. Explicit key: same key twice -> one note, first content wins,
                #    same id returned both times.
                k = {"title": "Retme", "content": "first", "idempotency_key": "op-42"}
                r4 = await session.call_tool("add_note", k)
                r5 = await session.call_tool(
                    "add_note", {"title": "Retme", "content": "SECOND try", "idempotency_key": "op-42"}
                )
                id4 = _text(r4).rsplit(" ", 1)[-1].rstrip(".")
                assert id4 in _text(r5), f"same key returned a different note: {_text(r5)}"
                assert len(await _notes(session)) == start + 3, "reused key created a second note"
                stored = [n for n in await _notes(session) if n.get("idempotency_key") == "op-42"]
                assert len(stored) == 1 and stored[0]["content"] == "first", stored

        # 4. Survives a restart: the keyed note is still deduped by a fresh process.
        async with stdio_client(_params(data_root)) as (read, write):
            async with ClientSession(read, write, list_roots_callback=_roots_cb(data_root)) as session:
                await session.initialize()
                before = len(await _notes(session))
                r6 = await session.call_tool(
                    "add_note", {"title": "x", "content": "y", "idempotency_key": "op-42"}
                )
                assert "already stored" in _text(r6), _text(r6)
                assert len(await _notes(session)) == before

        print("PASS: add_note dedups on idempotency_key and on identical title+content.")


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except Exception as exc:  # noqa: BLE001 - top-level test entry point
        print(f"FAIL: {exc}", file=sys.stderr)
        sys.exit(1)
